File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <queue>

using namespace std;

int BOARD[101][101];
int visited[101][101];
int ret = 0;
int n,m;

const int dy[4] = {1,-1,0,0};
const int dx[4] = {0,0,1,-1};

void bfs(int y, int x){
    queue<pair<int,int>> q;
    q.push({y,x});
    visited[y][x] = 1;
    while(q.size()){
        int y = q.front().first;
        int x = q.front().second;
        q.pop();
        if(y == n-1 && x == m-1)
            return;
        
        for(int dir = 0; dir < 4; dir++){
            int ny = y + dy[dir];
            int nx = x + dx[dir];
            
            if(ny < 0 || ny >= n || nx < 0 || nx >= m || visited[ny][nx] || !BOARD[ny][nx])
                continue;
            visited[ny][nx] = visited[y][x] + 1;
            q.push({ny,nx});
        }
    }
}

int solution(vector<vector<int>> maps)
{
    int answer = 0;
    n = maps.size();
    m = maps[0].size();
    for(int y = 0; y < maps.size(); y++){
        for(int x = 0; x < maps[y].size(); x++){
            if(maps[y][x])
                BOARD[y][x] = 1;
        }
    }
    bfs(0,0);
    answer = visited[n-1][m-1];
    if(answer == 0)
        return -1;
    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    private static int[] dy = {-1,1,0,0};
    private static int[] dx = {0,0,-1,1};
    static int[][] dist;
    static int N, M;
    
    public int solution(int[][] maps) {
        int answer = 0;
        N = maps.length;
        M = maps[0].length;
        dist = new int[N][M];
        BFS(maps);
        if(dist[N-1][M-1] == 0)
            return -1;
        return dist[N-1][M-1];
    }
    
    public void BFS(int[][] maps){
        Queue<Pos> q = new LinkedList<>();
        q.offer(new Pos(0,0));
        dist[0][0] = 1;
        while(!q.isEmpty()){
            Pos cur = q.poll();
            
            for(int dir = 0; dir < 4; dir++){
                int ny = cur.y + dy[dir];
                int nx = cur.x + dx[dir];
                
                if(ny < 0 || ny >= N || nx < 0 || nx >= M || dist[ny][nx] != 0 || maps[ny][nx] == 0)
                    continue;
                q.offer(new Pos(ny,nx));
                dist[ny][nx] = dist[cur.y][cur.x] + 1;
            }
        }
    }
}

class Pos{
    int y;
    int x;
    
    public Pos(int y, int x){
        this.y = y;
        this.x = x;
    }
}